### PR TITLE
Refactor tracing to improve maintainability

### PR DESCRIPTION
**Pull request by abaire**
This is a squash of a number of commits from
https://github.com/JayFoxRox/nv2a-trace/tree/tracing-refactor with permission from @JayFoxRox. I have a few improvements that were far easier to make with his refactored branch and would like to get his changes merged. For clarity, I'm happy to take on responsibility for any fixes or followups on these changes.

On my 1.0 xbox w/ xbdm, the current `master` code fails to retrieve anything whereas this refactor works correctly. I believe @antangelo may have had a similar experience as I was pointed at JFR's branch via their input.

Improve debug output in parseCommand (fixme: review)
Experiments with new tracing code
Move fifo kick to Xbox for stability
Disable tiling and zcompression at startup
SPLITME: surface cleanup and AA
FIXME dump PFB (depends on surface code cleanup)
Move texture dumper back into tracer (we need access to the info to generate memdumps)
VP RDI dumping
Improve framebuffer addressing using base
Record initial state
More explicit error messages

### CODE REVIEW

**Comment by abaire** [comment]
@mborgerson I'm curious as to your thoughts on how to proceed with this PR. It's quite large and since I can't get the current `master` to work on my machine, I don't have any proof that it doesn't cause regressions.

Do you happen to know if the current `master` works or is it possible that it has bitrotted due to changes in `xboxpy` that were never reflected here? If it is broken for folks other than myself, I wonder if we can merge this wholesale and improve iteratively? I have a couple fixes/improvements that JFR already merged https://github.com/JayFoxRox/nv2a-trace/commits/tracing-refactor and a [WIP commit](https://github.com/abaire/nv2a-trace/commit/c2a97043e44df4bce2d9c9e28d61e27df8349304) to add simplified nv2a logs for easier comparative work. I plan to also investigate the slow framebuffer captures I've been experiencing, and I believe @antangelo was going to look into some instability he was experiencing.

Alternatively, maybe we should stand up a "v2" branch (or even repo) until there's sufficient burn in time through use by xemu/elsewhere contributors to be confident that it doesn't introduce issues that would block active users of `master`?

**Comment by antangelo** [comment]
I've used the current master branch in the past with mostly successful results. I've noticed that JFR's tracing refactor branch (and this one by extension) omit some of the pgraph register dumping that master does (set [here](https://github.com/XboxDev/nv2a-trace/blob/master/Trace.py#L69) as an example), but this can be trivially re-enabled using the new method hooks as needed.

I haven't gotten successful texture or surface dumps from master in most cases, whereas JFR's branch has been able to dump the same textures and surfaces successfully.

The root cause of the stability issues I've been experiencing may be due to xboxpy, but I haven't looked too closely into them yet. In any case, the stability is about the same in my experience between master and JFR's branch. I'll test this branch specifically when I have time.

**Comment by mborgerson** [comment]
> @mborgerson I'm curious as to your thoughts on how to proceed with this PR. It's quite large and since I can't get the current `master` to work on my machine, I don't have any proof that it doesn't cause regressions. Do you happen to know if the current `master` works or is it possible that it has bitrotted due to changes in `xboxpy` that were never reflected here? If it is broken for folks other than myself, I wonder if we can merge this wholesale and improve iteratively?

Yeah, I'm perfectly okay with iterating here. This repo is still new in terms of maturity so we have some flexibility to prioritize function.

> I have a couple fixes/improvements that JFR already merged https://github.com/JayFoxRox/nv2a-trace/commits/tracing-refactor and a [WIP commit](https://github.com/abaire/nv2a-trace/commit/c2a97043e44df4bce2d9c9e28d61e27df8349304) to add simplified nv2a logs for easier comparative work. I plan to also investigate the slow framebuffer captures I've been experiencing, and I believe @antangelo was going to look into some instability he was experiencing.

Great

> Alternatively, maybe we should stand up a "v2" branch (or even repo) until there's sufficient burn in time through use by xemu/elsewhere contributors to be confident that it doesn't introduce issues that would block active users of `master`?

Let's not worry about it. I'm not aware of anyone else using this right now, so let's move forward with improvements.

One thing we should address in the immediate future is project license. Neither the original code nor this new code has a license attached to it, so we should ask @JayFoxRox how he wishes to license this code.

**Comment by MasonT8198** [comment]
@abaire I can't seem to get latest master to work properly for me however using your branch works perfectly

**Comment by JayFoxRox** [comment]
> so we should ask @JayFoxRox how he wishes to license this code.

Feel free to use it under MIT License, so it's compatible with all Python libs.

**Comment by abaire** [comment]
> > so we should ask @JayFoxRox how he wishes to license this code.
> 
> Feel free to use it under MIT License, so it's compatible with all Python libs.

Thanks! Added a `LICENSE.md` w/ the MIT license to this PR so it's clear for the future.

File: LICENSE.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+Copyright (c) 2012-2022 Scott Chacon and others
```
Comment (JayFoxRox) [inline LICENSE.md:1]: This is misleading. I wrote the majority of this, so if we name any individual it should probably be me.

Generally I'm more fond of listing an org like "XboxDev Maintainers" (although that adds some legal ambiguity) or *all* contributors explicitly.

Comment (JayFoxRox) [inline LICENSE.md:1]: I'd also prefer to not have my e-mail listed in the commit title. GitHub will only list e-mails in commits so web-crawlers don't see them.

Comment (abaire) [reply LICENSE.md]: Total oversight on my part, I copied the license text from a sample on GH. Fixed!

I went with XboxDev maintainers, as I am dubious that the maintenance overhead of keeping an explicit list of contributors is worthwhile and using "and others" seems just as ambiguous as the blanket maintainers. Happy to change if you'd prefer.

Comment (abaire) [reply LICENSE.md]: Makes sense, amended the message to just use your name.

Comment (JayFoxRox) [reply LICENSE.md]: Looks okay now; the README still mentions "All Rights Reserved" I believe - that'd be incorrect now.

Comment (abaire) [reply LICENSE.md]: Fixed, and lowercased "maintainers" in `LICENSE.md` to align with use in `README`

**Comment by mborgerson** [comment]
With the understanding that this code will be iteratively improved, I've given a quick pass and have no major objections. There are several code quality improvements I'd like to see done here, and we can improve it in follow ups. Big thanks to @JayFoxRox for sharing the work done here and thanks to @abaire for working on it; this will be an invaluable tool!

**Pull request merged**